video: selection api (crop/compose) addition (api/shell/sample) + dcmipp implementation

avolmat-st opened the pull request:
This PR is still a WIP but should already give a good idea of the proposed SELECTION API allowing to the get capabilities of crop/compose as well as control it.
This is based on the Linux V4L2 SELECTION API (available in BSD-Clause): https://docs.kernel.org/6.6/userspace-api/media/v4l/selection-api.html

In order to give a better idea, I've also implemented the shell part as well as DCMIPP driver part (ongoing since only pixel pipes done for the time being, still need a few more code to also add the dump pipe support).

Basically, the setting a crop means that only part of the input is taken into account. The selected crop becomes the default compose size.
Setting a compose size means that the area selected via the crop would get modified (down or up) in order to lead to the rectangle defined by the compose. This lead to the output format at the end.

This leads for example to stuff as what can be seen in dcmipp modifications which is:
  - setting a crop lead to also setting the compose to the same rectangle and the output format width/height as well
  - setting a compose does not impact the crop but compose width/height also become the output format width/height
  - in such condition, set_fmt is mainly used to set the pixelformat, and the driver needs to check that the width/height of set_fmt do match with compose width/height.

In order to give a better idea I've modified the capture sample by adding new CONFIG_ in order to set the crop area and the sample app will, if available in the driver try to set the crop and do the compose to match with the requested output format.

Example of shell output, using DCMIPP driver on the STM32N6:

1. get output format upon start
```
uart:~$ video format endpoint@1 out
Current format: 'pRAA' 2592x1944 (0 bytes per frame)
min vbuf count: 1
min line count: 4294967295
max line count: 4294967295
	[0]: 'pRAA' (10 bits per pixel)
		Size: Discrete: 2592x1944 (6298560 bytes per frame)
			Interval: Discrete 33 ms (30 FPS)
```
2. Get current crop and apply a new one
```
uart:~$ video selection endpoint@1 out crop
	selection target: crop: (0,0)/2592x1944
uart:~$ video selection endpoint@1 out crop (10,10)/2400x1440
	selection target: crop: (10,10)/2400x1440
uart:~$ video selection endpoint@1 out crop
	selection target: crop: (10,10)/2400x1440
uart:~$ video format endpoint@1 out
Current format: 'pRAA' 2400x1440 (4331520 bytes per frame)
min vbuf count: 1
min line count: 4294967295
max line count: 4294967295
	[0]: 'pRAA' (10 bits per pixel)
		Size: Discrete: 2592x1944 (6298560 bytes per frame)
			Interval: Discrete 33 ms (30 FPS)
```
3. Set a compose to perform downscale
```
uart:~$ video selection endpoint@1 out compose
	selection target: compose: (10,10)/2400x1440
                                                     ^^^^^
                                                     this is actually wrong, a but within the DCMIPP code I'll fix it in the next push
uart:~$ video selection endpoint@1 out compose (0,0)/800x480
	selection target: compose: (0,0)/800x480
uart:~$ video selection endpoint@1 out compose
	selection target: compose: (0,0)/800x480
uart:~$ video format endpoint@1 out
Current format: 'pRAA' 800x480 (483840 bytes per frame)
min vbuf count: 1
min line count: 4294967295
max line count: 4294967295
	[0]: 'pRAA' (10 bits per pixel)
		Size: Discrete: 2592x1944 (6298560 bytes per frame)
			Interval: Discrete 33 ms (30 FPS)
```
Upon performing crop / compose the DCMIPP will now generate a 800x480 frame which is a downscale from the a rectangle of 2400x1440 starting at offset 10,10 of an original sensor frame of 2592x1944

Commit:
video: api: addition of video_set/get_selection APIs

Addition of selection APIs (set/get) in order to be able
to control the crop and compose of a video device.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: stm32: dcmipp: add set/get_selection handling

Add possibility to perform crop on all pipes and compose (downscale) on
pixel pipes (endpoint #1 and endpoint #2).
Rework the code in order to move the downscale control from
the set_fmt into the set_selection (compose).

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: shell: addition of video selection support

Add shell handling in order to interact with drivers
via the set/get_selection APIs allowing to get/set
crop / compose and get HW capabilities such as
crop bound / compose bound or native size.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
samples: video: capture: add crop/compose support

Demonstrate the crop/compose API by introducing 4 new
CONFIG options in order to define the crop area.
Moreover, if the selection API is available and if
the targetted size is different from the current crop
size, then try to apply a compose in order to reach
the targetted format size.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

josuah reviewed (commented):
Thank you for introducing this API! This will be convenient for image sensors as well, which I suppose will use the same API for their cropping region since it is not 

I asked a few questions but it seems like being ready. Convenient to have V4L2 as a guideline!

I will give another cycle after your feedback, to make sure I understand what I review. Thank you!

josuah commented:
For reference, the linux-side equivalent:
- https://elixir.bootlin.com/linux/v6.15-rc6/source/include/uapi/linux/v4l2-common.h#L22
- https://elixir.bootlin.com/linux/v6.15-rc6/source/include/uapi/linux/videodev2.h#L1360

avolmat-st commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -139,6 +143,26 @@ int main(void)
+	sel.rect.top = CONFIG_VIDEO_SOURCE_CROP_TOP;
+	sel.rect.width = CONFIG_VIDEO_SOURCE_CROP_WIDTH;
+	sel.rect.height = CONFIG_VIDEO_SOURCE_CROP_HEIGHT;
+	if (video_set_selection(video_dev, &sel)) {
+		LOG_ERR("Unable to set selection crop");
+		return 0;
```
I'd prefer not change this. The whole sample app is actually returning 0 on error after sending a LOG_ERR.

avolmat-st commented on drivers/video/video_stm32_dcmipp.c:
```diff
@@ -1183,6 +1231,99 @@ static int stm32_dcmipp_enum_frmival(const struct device *dev, struct video_frmi
+			LOG_ERR("Crop width/height (%d x %d) not part of original frame",
+				sel->rect.width, sel->rect.height);
+			return -EINVAL;
+		}
+		pipe->crop = sel->rect;
+		pipe->compose = pipe->crop;
+		pipe->fmt.width = sel->rect.width;
+		pipe->fmt.height = sel->rect.height;
+		stm32_dcmipp_compute_fmt_pitch(pipe->id, &pipe->fmt);
```
Indeed, for the mutex added in this function to protect crop/compose/fmt variable which can be used in other places such as stream_on for example.

avolmat-st commented on drivers/video/video_shell.c:
```diff
@@ -1045,6 +1045,156 @@ static void complete_video_format_dev(size_t idx, struct shell_static_entry *ent
+		*sel_target = VIDEO_SEL_TGT_COMPOSE_BOUND;
+	} else {
+		shell_error(sh,
+			    "Target must be 'crop/crop_bound/native_size/compose or compose_bound");
```
Modify as proposed by @josuah.

Commit:
video: shell: correct incorrect error return value in cmd_video_format

Correct cmd_video_format return value in case of video_shell_parse_in_out
returns a negative value.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

avolmat-st commented on include/zephyr/drivers/video.h:
```diff
@@ -756,6 +818,69 @@ static inline int video_set_signal(const struct device *dev, struct k_poll_signa
+	__ASSERT_NO_MSG(sel != NULL);
+
+	api = (const struct video_driver_api *)dev->api;
+	if (api->set_selection == NULL) {
```
Oups, bad copy & paste

avolmat-st commented on include/zephyr/drivers/video.h:
```diff
@@ -237,6 +238,57 @@ enum video_signal_result {
+ * Used for crop/compose and possibly within drivers as well
+ */
+struct video_rect {
+	/** left offset of selection rectangle */
+	uint32_t left;
+	/** top offset of selection rectangle */
+	uint32_t top;
+	/** width of selection rectangle */
+	uint32_t width;
+	/** height of selection rectangle */
+	uint32_t height;
```
Yeah I also think that uint16_t would be enough. There could be some cases when sensor send multiple frames in one, so ending up with very big height, but still, 65535 sounds like already quite big I think. But in this case the whole video API would have to be changed to uint16_t instead of uint32_t.

josuah replied:
That makes sense, that would be for another PR.
Also yes, line scanners sensor could have readouts of kilometers long maybe... :)

avolmat-st commented on drivers/video/video_stm32_dcmipp.c:
```diff
@@ -1183,6 +1231,99 @@ static int stm32_dcmipp_enum_frmival(const struct device *dev, struct video_frmi
+		}
+		pipe->crop = sel->rect;
+		pipe->compose = pipe->crop;
+		pipe->fmt.width = sel->rect.width;
+		pipe->fmt.height = sel->rect.height;
```
Yes, you can understand this as "compose" describe how the cropped rectangle (defined by "crop") is going to be output. So, as soon as you change the crop, the compose doesn't make anymore sense. So, you first need to select the area you are going to work on (aka doing the crop), then say how want this to be composed (compose).
There is really need to always select the crop either, if the default is fine, then no need to change it, but it is important to understand that:

1. performing a crop will reset the compose to the crop rectangle (aka composing 1:1 in a sense)
2. performing a compose will set the output format (obviously since, if you say you want to up-scale or downscale, the output frame size will definitively change).

So basically, you can consider  CROP -> COMPOSE -> FMT
Changing one will reset the next stage. This is of course not true in reverse. Even if you change the COMPOSE, the CROP won't change.

josuah replied:
Thank you, I think I missed a few lines from your first PR message that explained this.
This makes sense now!

avolmat-st commented on drivers/video/video_shell.c:
```diff
@@ -1045,6 +1045,156 @@ static void complete_video_format_dev(size_t idx, struct shell_static_entry *ent
+	}
+	end_size++;
+
+	if (*end_size != '/') {
+		shell_error(sh,
+			    "Invalid rectangle format (<left>,<top>)/<width>x<height> parameter %s",
+			    arg_rect);
+		return -EINVAL;
+	}
+	end_size++;
```
Yep indeed, I change to <left> <top> <width>x<height>

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -756,6 +818,69 @@ static inline int video_set_signal(const struct device *dev, struct k_poll_signa
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param sel Pointer to a video selection structure
+ *
```
[non-blocking]
```suggestion
 * Some targets are inter-dependents. For instance, setting a @ref VIDEO_SEL_TGT_CROP will
 * reset @ref VIDEO_SEL_TGT_COMPOSE to the same size when both are supported.
```
Here is some minimum hint for users who can't figure out why their firmware work only when calling functions in different order leads to different results.

josuah commented:
Maybe the challenge arose from `STM32_DCMIPP_MAX_PIPE_SCALE_FACTOR`: it is not an absolute limitation, but a "percentage of the crop input size" limitation...

The way this was solved for frame size+interval combo was to first set the size, then query the interval.

The equivalent would be to set the crop size first `VIDEO_SEL_TGT_CROP`, and then query the supported compose size `VIDEO_SEL_TGT_COMPOSE_BOUND`.

This would be the compose size returned for the current crop settings, taking `STM32_DCMIPP_MAX_PIPE_SCALE_FACTOR` into account.

Does that make sense?

josuah commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -147,6 +166,31 @@ int main(void)
+	 */
+	sel.target = VIDEO_SEL_TGT_CROP;
+	err = video_get_selection(video_dev, &sel);
+	if (err == 0) {
+		if (sel.rect.width != fmt.width || sel.rect.height != fmt.height) {
+			sel.target = VIDEO_SEL_TGT_COMPOSE;
+			sel.rect.left = 0;
+			sel.rect.top = 0;
+			sel.rect.width = fmt.width;
+			sel.rect.height = fmt.height;
+			err = video_set_selection(video_dev, &sel);
+			if (err != -ENOSYS && err < 0) {
+				LOG_ERR("Unable to set selection compose");
+				return 0;
+			}
+		}
+	} else if (err != -ENOSYS && err < 0) {
+		LOG_ERR("Unable to get selection crop");
+		return 0;
+	}
```
[non-blocking]
Proposal for making the error-handling idioms appear, just a minor nit:
```c
err = video_get_selection();            //
if (err < 0 && err != -ENOSYS) {        // well-known C error handling idiom
    LOG_ERR("That did not work");       //
    return 0;                           //
}                                       //

if (err == 0) {
    err = video_set_selection();        //
    if (err < 0 && err != -ENOSYS) {    // well-known C error handling idiom
        LOG_ERR("That did not work");   //
        return 0;                       //
    }                                   //
}
```

josuah replied:
Sometimes what is logical to write is not the same as what is logical to read. I got myself tricked in this exactly earlier.

avolmat-st commented:
> Thank you I understand better and agree.
> 
> If the information needed to make a choice is in the driver, then the driver has to expose that (is it even possible?) or choose and tell what it chose.

Yes, I think this depends on the HW so hard for an application to know that. That's why I chose to have the driver update the requested crop/compose and let the application know what has been set. But this implie that application check again after getting back the value and take some additional action if the applied setting aren't good for the application.

Such behavior is ok I think for a Linux application, but I am not sure if this match well what constraint of a Zephyr application, hence I am still wondering if this should be just "check if this can be applied to the HW and simply reject it if not".

josuah reviewed (dismissed):
Nothing blocking left I could find! Thank you for the modifications!

avolmat-st commented on drivers/video/video_stm32_dcmipp.c:
```diff
@@ -1183,6 +1272,172 @@ static int stm32_dcmipp_enum_frmival(const struct device *dev, struct video_frmi
+			goto out;
+		}
+
+		if (sel->rect.left != 0) {
+			sel->rect.left = 0;
+		}
+		if (sel->rect.top != 0) {
+			sel->rect.top = 0;
+		}
```
it set left and/or top to 0 if it not 0. if it is already 0 it won't do anything. But yes, always setting left and top to 0 would make the code easier to read/understand I think.

ngphibang commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -139,6 +142,22 @@ int main(void)
+		sel.rect.left, sel.rect.top, sel.rect.width, sel.rect.height);
+#endif
+
+#if CONFIG_VIDEO_FRAME_HEIGHT || CONFIG_VIDEO_FRAME_WIDTH
```
nit: In a function, it is recommended to use `if (IS_ENABLED())` but it is currently not consistent across the sample

josuah replied:
This would also solve the CI issue.

avolmat-st replied:
Done. Added an additionnal prior commit to use if (IS_ENABLED(CONFIG_VIDEO_FRAME_HEIGHT) ... instead of preprocessor.

ngphibang commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -147,6 +166,31 @@ int main(void)
+	sel.target = VIDEO_SEL_TGT_CROP;
+	err = video_get_selection(video_dev, &sel);
+	if (err == 0) {
+		if (sel.rect.width != fmt.width || sel.rect.height != fmt.height) {
+			sel.target = VIDEO_SEL_TGT_COMPOSE;
+			sel.rect.left = 0;
+			sel.rect.top = 0;
+			sel.rect.width = fmt.width;
+			sel.rect.height = fmt.height;
+			err = video_set_selection(video_dev, &sel);
+			if (err != -ENOSYS && err < 0) {
+				LOG_ERR("Unable to set selection compose");
+				return 0;
+			}
+		}
```
I think we should not impose this constraint.

- The format & crop rectangle can be mismatch. The framework does not impose any constraint on this and it's up to the implementation in each driver, I think. If the HW supports scaling and the driver (wants) to implement this support, so it's fine.
- The 2nd reason is that get/set_selection is per subdev, not propagated, so here we just query the 1st parent device ...

To demonstrate the API, I think it's fine to just print the information.

ngphibang reviewed (commented):
LGTM except the constraint imposed on the sample where I think it should be up to the driver to decide.

Thanks for adding this API. The selection API looks simple enough but the complexity is left in the driver implementation and user's usage, especially its interaction with get/set_format(), e.g. what if format is set first, selection is set first, size does not mismatch, etc. It is left to the driver to decide to tightly couple with set_fmt() or simply return error. And users should always read back their original settings to decide ...

josuah commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -95,6 +95,9 @@ int main(void)
 	struct video_frmival frmival;
 	struct video_frmival_enum fie;
 	enum video_buf_type type = VIDEO_BUF_TYPE_OUTPUT;
+	struct video_selection sel = {
+		.type = VIDEO_BUF_TYPE_OUTPUT,
+	};
```
- Solved by using `if (IS_ENABLED())`

GeorgeCGV commented on drivers/video/video_stm32_dcmipp.c:
```diff
@@ -1183,6 +1272,173 @@ static int stm32_dcmipp_enum_frmival(const struct device *dev, struct video_frmi
+	case VIDEO_SEL_TGT_COMPOSE:
+		sel->rect = pipe->compose;
+		break;
+	case VIDEO_SEL_TGT_CROP_BOUND:
```
Minor:
```suggestion
	case VIDEO_SEL_TGT_CROP_BOUND:
		__fallthrough;
```
